### ISSUE TITLE
Module proposal: win_wininet (configure IE proxy settings)

### ISSUE THREAD

**Issue by johnboy2** [ansible/ansible#22663]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME
win_wininet (proposed name)

##### ANSIBLE VERSION

```
2.2.1.0
```

##### SUMMARY
It hard/non-obvious to configure the ansible_user's "Internet Options" proxy configuration on Windows. This proposal suggests adding a new module to do that.

The potential benefit is that various other downloaders could then be made to work *implicitly* using wininet.dll. For example, .Net's System.Net.WebClient, or the "win_chocolatey" module.

##### OTHER NOTES

There are a number of **wrong** proposals online for configuring the IE proxy settings:
* Just set some specific registry keys (only works if the user account has had its IE profile initialized previously by running IE *at least once* -- which isn't helpful at all for automation)
* Set some registry keys and then "poke" wininet.dll to force a reload (same issue as last point)
* Use "netsh winhttp set proxy ..." (fundamentally misunderstands the difference between wininet and winhttp, which are different libraries used by different things in Windows)

A better solution is to initialize IE via wininet.dll itself. This can be done in PowerShell through inline C# with data marshaling for options content. Such a solution should be fairly easy for Ansible to deliver/manage.

I have a proof-of-concept for this as a binary written in C, which I have been using so far by first uploading the binary, executing it, and then cleaning it up, in multiple, separate Ansible tasks. (I can post the relevant source as soon as I get it cleared by my employer.)

**Comment by dagwieers** [comment]
A module for this is acceptable. However win_wininet really looks weird. I wonder if another (more descriptive) name makes more sense, like **win_internet_options** or **win_proxy**.

I would say, go for it !

**Comment by johnboy2** [comment]
@dagwieers good point RE name, though **win_proxy** might too ambiguous (wininet vs winhttp); **win_internet_options** seems good though, because of its similarity to the control panel name for wininet. Would also make it more generalizable for the future (eg to add some non-proxy options in the future).

**Comment by johnboy2** [comment]
PoC code now available at https://gist.github.com/johnboy2/a862060b386d354f488ca2c6049e9105. 

Note that I am merely "putting this out there" in case it is useful to others, or if somebody wants to take on modularizing this for Ansible. (I won't likely have time to attempt doing that myself any time soon -- certainly not within the next few months; which is why I'm making this available.)

**Comment by jborean93** [comment]
I'm all for having a module for this and I can't claim to know the specifics but I'm unsure as to whether we need to use some PInvoke to the `wininet.dll` as I've gotten this to work on Server 2016. If I run the following task;

```
- name: Set the default profile proxy settings
  win_regedit:
    path: HKLM:\ANSIBLE\Software\Microsoft\Windows\CurrentVersion\Internet Settings
    name: '{{ item.name }}'
    data: '{{ item.data }}'
    type: '{{ item.type }}'
    state: present
    hive: C:\Users\Default\NTUSER.dat
  with_items:
  - name: ProxyEnable
    data: 1
    type: dword
  - name: ProxyServer
    data: proxy.example.org:8080
    type: string
```

Every new user profile that is created on that host will have the proxy host enabled and set to `proxy.example.org:8080`. Any existing user would either have to delete their profile and log in again, or manually configure the settings for their account.

**Comment by johnboy2** [comment]
In my experience, registry-based proxy changes only work **after** wininet has gone through first-time initialization (ie by opening up IE or at least the IE control panel _at least once_). If you don't do that, those changes are meaningless and ignored.

The approach referenced by my gist above covers the case where no such "first run" has occurred.

**Comment by jborean93** [comment]
I finally got round to doing this, a PR can be found at https://github.com/ansible/ansible/pull/54631. It uses the WinHTTP/WinINet API directly to avoid this issue, just like the gist supplied. I did make it specific to WinHTTP (win_http_proxy) and WinINet (win_inet_proxy) and not generic. If we ever have the desire to manage IE settings then I feel that a separate module would be warranted. The code just for the WinINet proxy stuff was already large enough and I would hate to see how complex a WinINet settings module would be.
